Stop print timer with M105/M109

jbrazio opened the pull request:
This is a proposed fix for #3061.
Three hooks have been set in order to stop the print timer:
1. `M105`
2. `M109`
3. `disable_all_heaters()`

Note: `M105/M109` hooks only stop the print timer when the temperature of the last extruder is set to zero.

jbrazio committed:
Print timer now stops when it sees the last extruder temperature being shutdown

thinkyhead commented:
There's a very remote possibility that in a multi-extruder system, one extruder may finish its part and shut off, while the other continues. If the order of operations is COOL 1, HEAT 2 instead of HEAT 2, COOL 1, then the timer would be stopped by this code. I don't really think this is a problem. In any real, working print system, there should never be a time during a print when all heaters are off.

Should other commands, like `M81` also turn off the print timer?

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -3944,16 +3957,15 @@ inline void gcode_M105() {
  *       Rxxx Wait for extruder(s) to reach temperature. Waits when heating and cooling.
  */
 inline void gcode_M109() {
+  float temp;
```
Should be `float temp = 0;` because it's comparing with 0 below…

thinkyhead commented on Marlin/dogm_lcd_implementation.h:
```diff
@@ -306,7 +306,8 @@ static void lcd_implementation_status_screen() {
     if (print_job_start_ms != 0) {
-      uint16_t time = (millis() - print_job_start_ms) / 60000;
+      uint16_t time = ((print_job_stop_ms > print_job_start_ms)
+                       ? print_job_stop_ms : millis()) / 60000 - print_job_start_ms / 60000;
```
Might as well only divide by 60000 once…

``` cpp
uint16_t time = ((print_job_stop_ms > print_job_start_ms
                 ? print_job_stop_ms : millis()) - print_job_start_ms) / 60000;
```

thinkyhead commented on Marlin/ultralcd_implementation_hitachi_HD44780.h:
```diff
@@ -707,7 +707,8 @@ static void lcd_implementation_status_screen() {
     if (print_job_start_ms != 0) {
-      uint16_t time = millis() / 60000 - print_job_start_ms / 60000;
+      uint16_t time = ((print_job_stop_ms > print_job_start_ms)
+                       ? print_job_stop_ms : millis()) / 60000 - print_job_start_ms / 60000;
```
…same idea…

``` cpp
uint16_t time = ((print_job_stop_ms > print_job_start_ms
                 ? print_job_stop_ms : millis()) - print_job_start_ms) / 60000;
```

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -3944,16 +3957,15 @@ inline void gcode_M105() {
   no_wait_for_cooling = code_seen('S');
   if (no_wait_for_cooling || code_seen('R')) {
-    float temp = code_value();
+    temp = code_value();
```
Since `temp` is only ever set here, you can put this line here instead of below. Then the declaration of `temp` doesn't need to be moved up above either:

``` cpp
if (temp > 0) LCD_MESSAGEPGM(MSG_HEATING);
```

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -3833,6 +3833,19 @@ inline void gcode_M104() {
+  // Detect if a print job has finished.
+  // When the target temperature for all extruders is zero then we must have
+  // finished printing.
+  if( print_job_start_ms != 0 ) {
```
Minor quibble about the spacing. It should be `if (print_job_start_ms != 0)`. See https://github.com/MarlinFirmware/Marlin/wiki/Coding-Standards#spacing. (You can also leave out the `!= 0` if you feel like it and no one will complain. We aim not to be too pedantic in coding style if the meaning is fairly obvious.)

Blue-Marlin commented:
This clock is also used for sd-printing. Does it need changes there?
M109 is not a very reliable criterion for starting the job-timer. I never realized it's also working for USB-prints because I always wait only for `M190`. For the nozzles I use `M104`.
How can the clock be reset? Currently it seems to add the time of a second print?

This PR seems to improve the current behaviour. But the concept of starting/stopping it is questionable because most non-3d-printer uses of Marlin do not need any temperature. (Start-g-code? Buffer usage? State?)

jbrazio commented:
This code is indeed solving half of the problem. In fact the print timer code is messy, for instance it is only available if `SDCARD` is defined, and this flag isn't checked in half of the places where Marlin deals with the timer.

There's is also another issue I found when working in this PR but I'm preparing a post about it and will ask for your opinion there.

Nevertheless there are here some loose ends and a couple more hooks have to be introduced into this PR.

jbrazio committed:
Clean up the code a bit

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -3968,6 +3982,22 @@ inline void gcode_M109() {
+      print_job_stop_ms = millis();
+      LCD_MESSAGEPGM(WELCOME_MSG);
+    }
+  }
```
With this code repeating we might as well make a function and call it in both places:

``` cpp
bool finish_print_job() {
  if (print_job_start_ms) {
    for (int i = 0; i < EXTRUDERS; i++) if (degTargetHotend(i) > 0) return false;
    print_job_stop_ms = millis();
    return true;
  }
  return false;
}
```

In gcode_M104:

```
  finish_print_job();
```

In gcode_M109:

```
  if (finish_print_job()) LCD_MESSAGEPGM(WELCOME_MSG);
```

jbrazio replied:
Then I'll take the opportunity to rework the overall code a bit a implement a start_print_job() function also.

thinkyhead commented:
I agree with @Blue-Marlin that it's tricky to determine when a print starts or stops. But since we are already making these timers dependent on the wait-for-temperature routines, I don't think it's too far outside the scope to also turn off the timer in the case of all temperatures being turned off. We can also re-set the timer to zero in the case when all heaters are currently turned off, but one or more of them is turned on with a wait option.

Generally, host software will keep its own print timer anyway. This timer is especially for SD printing, but since it hooks into the wait-for-heating stuff it also comes in handy when you have a host that doesn't keep a job timer (if there are any).

jbrazio commented:
This is indeed purely a cosmetic change for users who have `SDCARD` enabled and print from a host as the timer will keep running after everything is finished.

jbrazio committed:
All print timer related activity now uses print_job_start(), print_job_timer() or print_job_stop()

jbrazio commented:
Fixed bug: If during the print the nozzle temp is changed by a `M109` the timer would reset itself.
Improvement: I've done a rework of print timer in Marlin, now all actions related with the timer have to go through three functions: `print_job_start()`, `print_job_timer()` or `print_job_stop()`; implemented all the feedback received in this thread.

AnHardt commented:
~~Currently M109 does not wait for cooling with 'R'.~~
Clock functions are ok with sd and temperature triggering.

AnHardt commented:
Sorry. Works. My target-temp was below EXTRUDE_MINTEMP/2. So that is also tested as working. ;-)

jbrazio commented:
Thanks !
@AnHardt  Do you confirm this PR is working properly ?

AnHardt commented:
Yes.
